reportsrv/reportsrv.go

```diff
@@ -0,0 +1,194 @@
+	logger *zap.Logger
+}
+
+type DB interface {
+	NoTxn(context.Context) db.Tx
+
+	AnalysisArtifactsForAnalysis(tx db.Tx, id pacta.AnalysisID) ([]*pacta.AnalysisArtifact, error)
+	Blobs(tx db.Tx, ids []pacta.BlobID) (map[pacta.BlobID]*pacta.Blob, error)
+}
+
+type Blob interface {
+	Scheme() blob.Scheme
+
+	ReadBlob(ctx context.Context, uri string) (io.ReadCloser, error)
+}
+
+func New(cfg *Config) (*Server, error) {
+	if err := cfg.validate(); err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+	return &Server{
+		db:     cfg.DB,
+		blob:   cfg.Blob,
+		logger: cfg.Logger,
+	}, nil
+}
+
+func (s *Server) verifyRequest(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// TODO(#12): Figure out how to share authz between the main pactasrv and this.
+		// Should be a fast follow, but large enough to be a separate change.
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (s *Server) RegisterHandlers(r chi.Router) {
+	r.Use(s.verifyRequest)
+	r.Get("/report/{analysis_id}/*", s.serveReport)
+}
+
+func (s *Server) serveReport(w http.ResponseWriter, r *http.Request) {
+	aID := pacta.AnalysisID(chi.URLParam(r, "analysis_id"))
+	if aID == "" {
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+	ctx := r.Context()
+
+	artifacts, err := s.db.AnalysisArtifactsForAnalysis(s.db.NoTxn(ctx), aID)
+	if err != nil {
+		s.logger.Error("failed to load artifacts for analysis", zap.String("analysis_id", string(aID)), zap.Error(err))
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	var blobIDs []pacta.BlobID
+	for _, a := range artifacts {
+		blobIDs = append(blobIDs, a.Blob.ID)
+	}
+
+	blobs, err := s.db.Blobs(s.db.NoTxn(ctx), blobIDs)
+	if err != nil {
+		s.logger.Error("failed to load blobs", zap.Strings("blob_ids", asStrs(blobIDs)), zap.Error(err))
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	subPath := strings.TrimPrefix(r.URL.Path, "/report/"+string(aID))
+	if strings.HasPrefix(subPath, "/") {
+		subPath = subPath[1:]
+	}
+	if subPath == "" {
+		subPath = "index.html"
+	}
+
+	for _, a := range artifacts {
+		// Container is just 'reports', we can ignore that.
+		b, ok := blobs[a.Blob.ID]
+		if !ok {
+			s.logger.Error("no blob loaded for blob ID", zap.String("analysis_artifact_id", string(a.ID)), zap.String("blob_id", string(a.Blob.ID)))
+			continue
+		}
+		uri := string(b.BlobURI)
+		_, path, ok := blob.SplitURI(s.blob.Scheme(), uri)
+		if !ok {
+			s.logger.Error("blob had invalid URI", zap.String("analysis_artifact_id", string(a.ID)), zap.String("blob_uri", uri))
+			continue
+		}
+		_, uriPath, ok := strings.Cut(path, "/")
+		if !ok {
+			s.logger.Error("path had no UUID prefix", zap.String("analysis_artifact_id", string(a.ID)), zap.String("blob_uri", uri), zap.String("blob_path", path))
+			continue
+		}
+
```

Comment:
One piece I don't love about this approach is that if N = number of assets in a given report, we read back N rows 2*N times to serve the report. I get that this is still linear in the # of DB calls made, it just seems really duplicative, since the returns for each of these DB calls is identical. I can't think of a simple way of sidestepping this property, but I don't love it.

Reply:
Yeah, I also didn't like that and considered doing a dumb in-mem cache of `analysis ID -> artifacts`, but the cardinality is low (e.g. 10-20 artifacts), and the loading speed isn't a problem. Can revisit if it becomes an issue.